src/libraries/Common/src/System/Security/Cryptography/CngPkcs8.cs

```diff
@@ -305,6 +350,56 @@ internal static unsafe Pkcs8Response ImportPkcs8PrivateKey(ReadOnlySpan<byte> so
             }
         }
 
+        // CNG cannot import a PrivateKeyInfo with the following criteria:
+        // 1. Is a EC key with explicitly encoded parameters
+        // 2. Is missing the PublicKey from ECPrivateKey.
+        // CNG can import an explicit EC PrivateKeyInfo if the PublicKey
+        // is present. CNG will also re-compute the public key from the
+        // private key if they do not much. To help CNG be able to import
+        // these keys, we re-write the PKCS8 to contain a zeroed PublicKey.
+        //
+        // If the PKCS8 key does not meet the above criteria, null is returned,
+        // signaling the original exception should be thrown.
+        private static unsafe AsnWriter? RewritePkcs8ECPrivateKeyWithZeroPublicKey(ReadOnlySpan<byte> source)
+        {
+            fixed (byte* ptr = &MemoryMarshal.GetReference(source))
+            {
+                using (MemoryManager<byte> manager = new PointerMemoryManager<byte>(ptr, source.Length))
+                {
+                    PrivateKeyInfoAsn privateKeyInfo = PrivateKeyInfoAsn.Decode(manager.Memory, AsnEncodingRules.BER);
+                    AlgorithmIdentifierAsn privateAlgorithm = privateKeyInfo.PrivateKeyAlgorithm;
+
+                    if (privateAlgorithm.Algorithm.Value != Oids.EcPublicKey)
```

Comment:
Should Oids.EcDiffieHellman work here for ECDH keys?

Reply:
@bartonjs I don't think so. Many other PKCS8 EC paths are limited to `EcPublicKey`:

https://github.com/dotnet/runtime/blob/6996928a6958f50a4cd811dfde9ea894473fc765/src/libraries/Common/src/System/Security/Cryptography/EccKeyFormatHelper.cs#L20-L23

Which is used here, for example:

https://github.com/dotnet/runtime/blob/6996928a6958f50a4cd811dfde9ea894473fc765/src/libraries/Common/src/System/Security/Cryptography/EccKeyFormatHelper.cs#L69-L70

Reply:
A Jeremy-from-the-past also thought of this:

https://github.com/dotnet/runtime/blob/07a5ee89b835a915ff3ac082ba3677fe39384a18/src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellman.cs#L14-L18

Reply:
Well, at least it means I'm consistently aware that there is a question to be asked :smile:.